### PR TITLE
Kenesbek HW07

### PR DESCRIPTION
**Pull request by serzhan-kenesbek**

### CODE REVIEW

**Comment by serzhan-kenesbek** [comment]
Most important files:
PBT/test/Test/Compiler.hs
PBT/test/Test/Interpreter.hs
PBT/test/Test.hs
7_2.txt

**Comment by serzhan-kenesbek** [comment]
Timestamps for my participation and dates:
04/10/2024 - Proof: Two screenshots of the Transcript
Screenshot 2024-11-15 at 00 13 49
Screenshot 2024-11-15 at 00 13 54

18/10/2024 - Proof: Two screenshots of timestamps
Screenshot 2024-11-15 at 00 10 00
Screenshot 2024-11-15 at 00 11 39

06/11/2024 - Proof: Screenshot of a timestamp
Screenshot 2024-11-14 at 23 59 55

**Comment by kajigor** [comment]
2 ok

File: PBT/test/Test/Compiler.hs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+-- Generator for numeric expressions
+genNumExpr :: Gen (E.Expr String)
+genNumExpr = E.Num <$> Gen.int (Range.linear (-1000) 1000)
+
+-- Generator for addition expressions
+genAddExpr :: Gen (E.Expr String)
+genAddExpr = do
+  left <- genNumExpr
+  right <- genNumExpr
+  return $ E.Plus left right
```
Comment (kajigor) [inline PBT/test/Test/Compiler.hs]: These are rather weak generators. Use `Gen.recursive` to generate random expressions and test them.

File: PBT/test/Test/Compiler.hs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+  case result of
+    Right finalState -> do
+      let stack = getStack finalState
+      stack === [evalExpr expr]
+    Left err -> do
+      footnote $ "Execution failed with error: " ++ show err
+      failure
+
+-- Helper to evaluate the expression
+evalExpr :: E.Expr v -> Int
```
Comment (kajigor) [inline PBT/test/Test/Compiler.hs]: You can implement the whole interpreter.

File: PBT/test/Test/Compiler.hs [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+  expr <- forAll genAddExpr
+  let program = compile expr
+  let result = execProgram program initialState
+  case result of
+    Right finalState -> do
+      let stack = getStack finalState
+      stack === [evalExpr expr]
+    Left err -> do
+      footnote $ "Execution failed with error: " ++ show err
+      failure
```
Comment (kajigor) [inline PBT/test/Test/Compiler.hs]: Generalize these properties and check that compiled expression evaluates to the same result.

**Review by kajigor — CHANGES_REQUESTED** [review]

**Comment by kajigor** [comment]
Ok

**Pull request closed without merging**